Document how to use the EFS provisioner (#409)
giantswarm/docs: src/content/guides/using-persistent-volumes-on-aws-with-efs/index.md

+++
title = "Using Persistent Volumes on AWS with EFS"
description = "Tutorial on how to use dynamically provisioned Persistent Volumes on a cluster running on Amazon Web Services with EFS"
date = "2020-02-06"
type = "page"
weight = 50
tags = ["tutorial"]
+++

# Using Persistent Volumes on AWS with EFS

If your cluster is running in the cloud on Amazon Web Services (AWS) the most common way to store data is using EBS volumes with the [dynamic provisioner](https://docs.giantswarm.io/guides/using-persistent-volumes-on-aws/). Sometimes EBS is not the optimal solution.

The advantages of using EFS over EBS are:

> - EFS data can be accessed from all AZ in the same region while EBS are tied to a single AZ.

Reviewer: ```suggestion
- EFS data can be accessed from all AZ in the same region while EBS is tied to a single AZ.
```

- EFS has the capability to mount the same Persistent Volume to multiple pods at the same time using the ReadWriteMany [access mode](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#access-modes).
- EFS will not hit the [AWS Instance Volume Limit](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/volume_limits.html) as it is a software mount and will avoid the [Impaired EBS](https://docs.giantswarm.io/guides/aws-impaired-volumes/) issue.
- EFS mount times are better than EBS.

> If you need to use EFS to provision volumes be careful with:

Reviewer 2: ```suggestion
If you need to use EFS to provision volumes, be advised:

> - All Kubernetes Persistent Volumes will be stored in the same EFS instance. You can deploy multiple provisioners per cluster, each having it's own storage-class and EFS instance.

Reviewer: ```suggestion
- All Kubernetes Persistent Volumes will be stored in the same EFS instance. You can deploy multiple provisioners per cluster, each having its own storage-class and EFS instance.
```

> - [EFS throughtput](https://docs.aws.amazon.com/efs/latest/ug/performance.html) need to be setup accordingly in order not to have performance issues. We only recommend Provisioned Throughput and if you need high performance you will need EBS.

Reviewer: ```suggestion
- [EFS throughtput](https://docs.aws.amazon.com/efs/latest/ug/performance.html) need to be set up accordingly in order not to have performance issues. We only recommend Provisioned Throughput and if you need high performance you will need EBS.
```

> - [EFS throughtput](https://docs.aws.amazon.com/efs/latest/ug/performance.html) need to be setup accordingly in order not to have performance issues. We only recommend Provisioned Throughput and if you need high performance you will need EBS.

Reviewer 2: ```suggestion
- [EFS throughtput](https://docs.aws.amazon.com/efs/latest/ug/performance.html) need to be set up accordingly in order to not have performance issues. We only recommend Provisioned Throughput, and if you need high performance, you will need EBS.
```

- EFS backups are done with [AWS Backup](https://aws.amazon.com/backup/) and it does not have the snapshot feature of EBS.

> - You cannot limit the amount of data stored in an EFS volume. The requested value in kubernetes is ignored.

Reviewer 2: ```suggestion 
You cannot limit the amount of data stored in an EFS volume. The requested value in Kubernetes is ignored.
```

> - EFS mount target are limited to 1 subnet per AZ. Each NodePool will create a different subnet per AZ, plan accordingly. 

Reviewer: ```suggestion
- EFS mount targets are limited to 1 subnet per AZ. Each NodePool will create a different subnet per AZ, plan accordingly. 
```

## Provision an EFS instance on AWS

> - Open the [AWS management](https://aws.amazon.com/console/) console in the account your cluster is located.

Reviewer: ```suggestion

- Open the [AWS management](https://aws.amazon.com/console/) console in the account your cluster is located.
```

- Select EFS from the services list.
- Create a new EFS mount and select the VPC where your cluster is located.
- Select the AZ with the subnets where your instances are located and the security-groups of the workers.

> - Choose the thoughput and performance mode, no file system policy or access points are needed.

Reviewer: ```suggestion
- Choose the throughput and performance mode, no file system policy or access points are needed.
```

- Create the instance and note down the EFS instance id.

## EFS Provisioner configuration file
Create a file on your local machine named "efs-provisioner.yaml" with the following content:
```yaml
global:
  deployEnv: production

efsProvisioner:
  efsFileSystemId: fs-90f935c8
  awsRegion: eu-central-1
  path: /
  provisionerName: giantswarm.io/aws-efs
  storageClass:
    name: efs
    isDefault: false
    gidAllocate:
      enabled: true
      gidMin: 40000
      gidMax: 50000
    reclaimPolicy: Delete
    mountOptions: []

rbac:
  create: true

podSecurityPolicy:
  enabled: true
```
You will need to adjust the **efsFileSystemId** and **awsRegion** to match the configured values of the EFS instance.

For additional configuration parameters see [documentation](https://github.com/kubernetes-incubator/external-storage/tree/master/aws/efs).

## Installing EFS Provisioner

> To install the provisioner you will need to follow this steps:

Reviewer: ```suggestion

To install the provisioner you will need to follow these steps:
```

- Access Happa and select the cluster you want to install the provisioner.
- Open the "Helm Stable" catalog

> - Write "efs-provisioner" in the search-bar.

Reviewer: ```suggestion
- Write "efs-provisioner" in the search bar.
```

- Select the "efs-provisioner" application and then click the Configure & Install button.
- Upload the efs-provisioner.yaml file created in the previous step.
- Finally, you can click the Install App button and it will be installed into your cluster.

> ## Using EFS Volumes

Reviewer: ```suggestion

## Using EFS Volumes
```

Your Kubernetes cluster will have a new Storage Class `efs` deployed, which you will have to reference when creating persistent volumes.

> In order to check that the stogare class exits you should see something similar to:

Reviewer: ```suggestion
In order to check that the storage class exits you should see something similar to:
```

> In order to check that the stogare class exits you should see something similar to:

Reviewer 2: ```suggestion
In order to check that the storage class exits, you should see something similar to:
```

```nohighlight
$ kubectl get storageclass
NAME            PROVISIONER             AGE
efs             giantswarm.io/aws-efs   23m
gp2 (default)   kubernetes.io/aws-ebs   5h31m
```

> In the following example you can see the annotation volume.beta.kubernetes.io/storage-class matches the "efs" Storage Class:

Reviewer: ```suggestion
In the following example you can see the annotation `volume.beta.kubernetes.io/storage-class` matches the "efs" Storage Class:
```

```yaml
kind: PersistentVolumeClaim
apiVersion: v1
metadata:
  name: test
  annotations:
    volume.beta.kubernetes.io/storage-class: "efs"
spec:
  accessModes:
    - ReadWriteMany
  resources:
    requests:
      storage: 100Mi
```

## Testing the new storage class

> You can create a file with the example above in a file called pvc_claim.yaml and instantiate with the following command:

Reviewer: ```suggestion

You can create a file with the example above in a file called pvc_claim.yaml and instantiate with the following command:
```

+++
title = "Using Persistent Volumes on AWS with EFS"
description = "Tutorial on how to use dynamically provisioned Persistent Volumes on a cluster running on Amazon Web Services with EFS"
date = "2020-02-06"
type = "page"
weight = 50
tags = ["tutorial"]
+++

# Using Persistent Volumes on AWS with EFS

If your cluster is running in the cloud on Amazon Web Services (AWS) the most common way to store data is using EBS volumes with the [dynamic provisioner](https://docs.giantswarm.io/guides/using-persistent-volumes-on-aws/). Sometimes EBS is not the optimal solution.

> 

Reviewer 3: - EBS: Please introduce an abbreviation on first use, like done with AWS.
- "Sometimes EBS is not the optimal solution." - is there a more positive way to put this, like "Compared to using EBS, EFS is a superior solution to this."

Reviewer 4: +1

Author: If your cluster is running in the cloud on Amazon Web Services (AWS) the most common way to store data is using Elastic Block Store volumes with the [dynamic provisioner](https://docs.giantswarm.io/guides/using-persistent-volumes-on-aws/). EBS uses hardware mounts which means it's slower to mount and there are [limits](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/volume_limits.html) on the number of volumes you can attach to the same node.

does that sound better?

Reviewer 3: Yes. And when you put (EBS) behind the long form, like in `data is using Elastic Block Store (EBS) volumes`, then it's even better and it is crystal clear what EBS stands for.

- EFS will not hit the [AWS Instance Volume Limit](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/volume_limits.html) as it is a software mount and will avoid the [Impaired EBS](https://docs.giantswarm.io/guides/aws-impaired-volumes/) issue.
- EFS mount times are better than EBS.

If you need to use EFS to provision volumes, be advised:
- All Kubernetes Persistent Volumes will be stored in the same EFS instance. You can deploy multiple provisioners per cluster, each having its own storage-class and EFS instance.
- [EFS throughtput](https://docs.aws.amazon.com/efs/latest/ug/performance.html) need to be set up accordingly in order to not have performance issues. We only recommend Provisioned Throughput, and if you need high performance you will need EBS.
- EFS backups are done with [AWS Backup](https://aws.amazon.com/backup/) and it does not have the snapshot feature of EBS.
- You cannot limit the amount of data stored in an EFS volume. The requested value in Kubernetes is ignored.
- EFS mount targets are limited to 1 subnet per Availability Zone. Each NodePool will create a different subnet per AZ, plan accordingly. 

## Provision an EFS instance on AWS

> 

Reviewer 3: - One line of intro for this section would be nice, like further down in the `Installing EFS Provisioner` section.
- Would the list make more sense as a numbered list instead of bullets? For step-by-step things that makes it more clear that the order is important.

Author: Added introduction

Author: Added introduction and numbered lists

- Open the [AWS management](https://aws.amazon.com/console/) console in the account your cluster is located.
- Select EFS from the services list.
- Create a new EFS mount and select the VPC where your cluster is located.
- Select the Availability Zone with the subnets where your instances are located and the security-groups of the workers.
- Choose the throughput and performance mode, no file system policy or access points are needed.
- Create the instance and note down the EFS instance id.

## EFS Provisioner configuration file
Create a file on your local machine named "efs-provisioner.yaml" with the following content:
```yaml
global:
  deployEnv: production

efsProvisioner:
  efsFileSystemId: fs-90f935c8
  awsRegion: eu-central-1
  path: /
  provisionerName: giantswarm.io/aws-efs
  storageClass:
    name: efs
    isDefault: false
    gidAllocate:
      enabled: true
      gidMin: 40000
      gidMax: 50000
    reclaimPolicy: Delete
    mountOptions: []

rbac:
  create: true

podSecurityPolicy:
  enabled: true

> ```

Reviewer 3: Is it clear here which values have to be adjusted by the user to match their case? When in doubt, add a line underneath, like "Adapt the awsRegion and ... to fit your needs"

Author: It's a couple lines down there, is this enough?

> You will need to populate the **efsFileSystemId** and **awsRegion** parameters to match the configured values of the EFS instance.

Reviewer 3: OK, thanks!

You will need to populate the **efsFileSystemId** and **awsRegion** parameters to match the configured values of the EFS instance.

For additional configuration parameters see [documentation](https://github.com/kubernetes-incubator/external-storage/tree/master/aws/efs).

## Installing EFS Provisioner

To install the provisioner you will need to follow these steps:
- Access Giant Swarm web UI and select the cluster on which you want to install the provisioner.

> - Open the "Helm Stable"

Reviewer 4: Sorry for deleting the word catalog after "Helm Stable" but I couldn't find a way to comment while editing.
That being said, there is no longer a "Helm Stable" catalog in the company jargon. It is either the "Giant Swarm App Catalog" or the "Playground App Catalog"

Author: No problem! If we are going to drop that Catalog then we might need to maintain this ourselves... which I am not sure we want to do for now.

I would like to keep that catalog if possible as there are some applications that work out of the box from upstream but I understand it can be confusing for customers

- Write "efs-provisioner" in the search bar.
- Select the "efs-provisioner" application and then click the Configure & Install button.
- Upload the efs-provisioner.yaml file created in the previous step.
- Finally, you can click the Install App button and it will be installed into your cluster.

> 

Reviewer 3: Again, maybe numerical ordering?

Author: changed

In order to check that the storage class exits, you should see something similar to:
```nohighlight
$ kubectl get storageclass
NAME            PROVISIONER             AGE
efs             giantswarm.io/aws-efs   23m
gp2 (default)   kubernetes.io/aws-ebs   5h31m
```

In the following example you can see the annotation `volume.beta.kubernetes.io/storage-class` matches the "efs" Storage Class:
```yaml
kind: PersistentVolumeClaim
apiVersion: v1
metadata:
  name: test
  annotations:
    volume.beta.kubernetes.io/storage-class: "efs"
spec:
  accessModes:
    - ReadWriteMany
  resources:
    requests:
      storage: 100Mi
```

## Testing the new storage class

You can create a file with the example above in a file called pvc_claim.yaml and instantiate with the following command:
```nohighlight
$ kubectl apply -f pvc_claim.yaml
```
In order to check the status of the volume we can check the status of the PVC:
```nohighlight
$ kubectl describe pvc test
NAME   STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS   AGE
test   Bound    pvc-0d7b988e-4eed-4cf7-918b-30964774fa13   100Mi        RWX            efs            5s

> ```

Reviewer 3: Is this really the output for the `describe` command? It looks more like a `get`.

Author: true!
